components/form/form.js

```diff
@@ -1,22 +1,53 @@
 import './form-errory-summary.js';
 import '../tooltip/tooltip.js';
 import '../link/link.js';
-import { findFormElements, getFormElementData, isCustomFormElement, isNativeFormElement } from './form-helper.js';
+import { findFormElements, flattenMap, getFormElementData, isCustomFormElement, isNativeFormElement } from './form-helper.js';
 import { html, LitElement } from 'lit-element/lit-element.js';
 import { getComposedActiveElement } from '../../helpers/focus.js';
 import { getUniqueId } from '../../helpers/uniqueId.js';
 import { LocalizeCoreElement } from '../../lang/localize-core-element.js';
 import { localizeFormElement } from './form-element-localize-helper.js';
 import { ValidationType } from './form-element-mixin.js';
 
+/**
+ * A component that can be used to build forms that validate and submit native and custom form elements.
+ * @slot - The native and custom form elements that participate in validation and submission
+ * @fires d2l-form-submit - Dispatched when the form is submitted
+ * @fires d2l-form-invalid - Dispatched when the form is submitted but one or more form elements failed validation
+ */
 class Form extends LocalizeCoreElement(LitElement) {
 
 	static get properties() {
 		return {
+			/**
+			 * The URL that processes the form submission
+			 */
 			action: { type: String },
+			/**
+			 * Causes the form to be submitted as an asynchronous XHR request rather than a native form submission.
+			 * If the form has nested forms then
```

Comment:
Looks like the comment got cut off. I'm curious about when you would want async XHR vs. not.

Reply:
The main use case for non-XHR would be if you want to display the response in some target (`_self`, `_blank`, `_parent`) like  with native form submission. Supporting this also means we can swap a native `form` out for a `d2l-form` and submission should still work the same.

XHR would be used whenever we want to be able to do some custom logic after submit rather.

That being said, I fully expect most modern use cases to have an event listener for `@d2l-form-submit`, call `preventDefault()` and submit via a Hypermedia action.

Reply:
Gotcha. Wouldn't this need a `target` attribute for the form then? Just wondering if we could get away without the attribute... like if they set the target then we use a native form, otherwise we XHR (assuming action is also set)?

Reply:
Yeah, it'll likely need a `target` attribute. I've been trying to think of a good way to allow options like that to be configured without having a bunch of attributes that only apply if other attributes are set but have yet to come up with a good solution.

Reply:
What about having `target`, `action`, `method` and `enctype` only apply to native form submission. If you want to do an XHR, add a listener for the `d2l-form-submit` event and do it yourself via fetch?

Because yeah otherwise... `target` doesn't make sense with XHRs and `action` is more of a form-y thing.

Reply:
The main reasons `d2l-form` knows how to do XHR submissions are:
1. When using nesting submission has to be `async` so if we'd then default to only firing a `d2l-form-submit` event and not actually submitting anything unless they have their own custom submit behavior. Its possible we're okay with that.
1. Building up a `FormData` object from a JSON object can be a pain if it includes things like files. If the form does the XHR submission then I can build it internally from the private native form with `new FormData(this._form)` rather than trying to piece it together manually.

If we aren't too concerned about those then we can get rid of it and just have native submission and custom submission if the have an event listener.